src/common/buffer.cc

```diff
@@ -2268,7 +2268,7 @@ int buffer::list::write_fd(int fd) const
   ssize_t bytes = 0;
 
   std::list<ptr>::const_iterator p = _buffers.begin();
-  while (p != _buffers.end()) {
+  while (p != _buffers.end() && iovlen != IOV_MAX) {
```

Comment:
Instead of checking this each iteration, you could just check the length of _buffers array beforehand.